hydradx/model/amm/arbitrage_agent.py

```diff
@@ -438,3 +438,208 @@ def calculate_profit(init_agent, agent, asset_map={}):
+                        lambda x: x[1] > 0 and x[0] != tkn_buy,  # all tokens we want to sell for which there is a price
+                        {
+                            x: test_ex.sell_spot(x, numeraire=tkn_buy)
+                            or (
+                                   test_ex.buy_spot(tkn_buy, numeraire=x)
+                                   if test_ex.buy_spot(tkn_buy, numeraire=x) > 0 else 0
+                               )
+                            for x in sell_tkns
+                        }.items()
+                    )
+                }
+                for tkn_sell in best_sell_tkns:
+                    sell_quantity, price = best_sell_tkns[tkn_sell]
+                    previous_tkn_sell = test_agent.holdings[tkn_sell]
+                    previous_tkn_buy = test_agent.holdings[tkn_buy]
+                    max_buy = test_ex.calculate_buy_from_sell(
+                        tkn_sell=tkn_sell,
+                        tkn_buy=tkn_buy,
+                        sell_quantity=sell_quantity
+                    )
+                    if max_buy <= 0:
+                        continue
+                    if max_buy <= buy_tkns[tkn_buy]:
+                        # buy as much as we can without going over sell_quantity
+                        test_ex.swap(test_agent, tkn_buy=tkn_buy, tkn_sell=tkn_sell, buy_quantity=max_buy)
+                        optimized_swaps.append({
+                            'exchange': ex_name,
+                            'trade': 'buy',
+                            'buy_asset': tkn_buy,
+                            'sell_asset': tkn_sell,
+                            'amount': max_buy
+                        })
+                        buy_tkns[tkn_buy] -= test_agent.holdings[tkn_buy] - previous_tkn_buy
+                        sell_tkns[tkn_sell] -= previous_tkn_sell - test_agent.holdings[tkn_sell]
+                    else:
+                        # buy enough to satisfy buy_quantity
+                        test_ex.swap(test_agent, tkn_buy=tkn_buy, tkn_sell=tkn_sell, buy_quantity=buy_tkns[tkn_buy])
+                        sell_tkns[tkn_sell] -= previous_tkn_sell - test_agent.holdings[tkn_sell]
+                        optimized_swaps.append({
+                            'exchange': ex_name,
+                            'trade': 'buy',
+                            'buy_asset': tkn_buy,
+                            'sell_asset': tkn_sell,
+                            'amount': buy_tkns[tkn_buy]
+                        })
+                        buy_tkns[tkn_buy] = 0
+                        break
+
+        if sum(buy_tkns.values()) > 0:
+            # try and sell everything remaining for USD, then use that to buy the remaining tokens
+            for tkn_sell in sell_tkns:
+                if sell_tkns[tkn_sell] > 0:
+                    test_ex.swap(
+                        agent=test_agent,
+                        tkn_buy=ex.stablecoin,
```

Comment:
I think instead of using the exchange's "stablecoin" for this, we should just designate a fallback asset for each exchange and have it separately configurable.

Reply:
I just set it to 'USD' for all cases. That's probably correct, and if it's not, easy to change.